### ISSUE TITLE
Beagle throws error when there is only one variant in input VCF

### ISSUE THREAD

**Issue by sbslee** [sbslee/pypgx#63]
A PyPGx user recently reported the following via email:

I have a couple of questions/issues while integrating with the latest version (1.5) of pypgx. I was hoping you could help me understand what's going on here.

Possible Beagle compatibility issue?

I'm attaching a sample vcf file to this email so you can check it out. It's GRCh37, ngs, CYP26A1
I get this error message when I'm running run_ngs_pipeline()

Captured stdout call

```
Saved VcfFrame[Imported] to: /tmp/tmpqbzlh8l4/out-CYP26A1/imported-variants.zip
```

Captured stderr call

```
Exception in thread "main" java.lang.IllegalArgumentException: Window has only one position: CHROM=10 POS=94833639
        at vcf.MarkerMap.meanSingleBaseGenDist(MarkerMap.java:98)
        at phase.FixedPhaseData.markerMap(FixedPhaseData.java:170)
        at phase.FixedPhaseData.<init>(FixedPhaseData.java:117)
        at main.Main.phaseAndImpute(Main.java:140)
        at main.Main.main(Main.java:110)
```

While looking up what this error means, I saw some folks talking about Beagle version being incompatible (input file format has changed since v4 and v5). Without really understanding what's going on =P, I tried replacing site-packages/pypgx/api/beagle.28Jun21.220.jar with another version (ex. Beagle v5.1 beagle.18May20.d20.jar) which seems to work around the issue... I'm hoping this means something to you =P

**Comment by sbslee** [comment]
This was my initial response:

I was able to reproduce the problem, which is weird given that this bug was supposedly fixed in a previous version of Beagle.

From the beagle5_release_notes file (https://faculty.washington.edu/browning/beagle/beagle5_release_notes):

```
Beagle 5.0 (15Sep18.8a0) release notes
=============================================
* Fixed a bug appearing when analysis window has only one target marker
```

FYI, PyPGx currently uses Beagle 5.2 (28Jun21.220). Since you mentioned that using an older version (e.g. Beagle 5.1 (18May20.d20)) fixed the issue -- and I can confirm it as well -- I suspect the same bug was re-introduced at some point after v5.1. The problem is, when I tried the latest version (Beagle 5.4 (05May22.33a)) the issue remained. I will try to contact Professor Browning to notify him of this bug. I will also think about ways to fix the bug in the PyPGx's side. Theoretically, it should be easy because when you have only one variant there is nothing to haplotype phase. But I will let you know.

**Comment by sbslee** [comment]
Below is my email conversation with Professor Browning who developed Beagle:

Thanks so much for the clarification, it all makes sense now :)


On Fri, Jun 3, 2022 at 12:40 AM Brian Browning <[browning@uw.edu](mailto:browning@uw.edu)> wrote:
Hi Steven,
You are correct that Beagle is *supposed* to raise this error whenever there is only one marker in a given window.  If a window contains a single marker and there are no missing genotypes, then replacing the unphased allele separator '/' with the phased allele separator '|' is a reasonable option.  However, if there are missing genotypes, changing the allele separator without imputing missing alleles would violate Beagle's guarantee that all output genotypes are phased and nonmissing.

At some point, I will attempt to implement a better solution to this problem, but it probably won't be this year because I have a full queue of other work that needs done.

Best regards,

Brian




On Wed, Jun 1, 2022 at 6:02 PM Seung-been Steven Lee <[sbstevenlee@gmail.com](mailto:sbstevenlee@gmail.com)> wrote:
Thanks for the reply!

Just to clarify, my main concern with the error was not about its message (i.e. "Window has only one position") but about the fact that the error seems to be raised conditionally depending on the Beagle version used.

As you can see from the attached standard error files, Beagle raised the error with v5.2 and v5.4, but not with v5.1.

Does this mean that v5.1 is in the wrong and that this bug was fixed in v5.2 (and v5.4) -- i.e. Beagle is *supposed* to raise this error whenever there is only one marker in a given window?

I'm asking because I'm currently using Beagle as part of my haplotype-based genotyping pipeline and I've really enjoyed v5.1 being able to handle input VCF with single variant (i.e. the most extreme case of single variant in given genetic window). However, if there's no plan to update Beagle to handle the only-one-marker-in-given-VCF situation, I can write a custom script to pseudo-phase these VCF files. Since there is only one variant in VCF, there's really nothing to haplotype phase to begin with ('0/1' can be changed to '0|1' without any statistical analysis).

So could you please let me know whether or not you have plan to update Beagle so that it can handle the only-one-marker-in-given-VCF situation?

Thanks.

Steven

On Wed, Jun 1, 2022 at 11:21 PM Brian Browning <[browning@uw.edu](mailto:browning@uw.edu)> wrote:
Dear Steven,
I apologize for the delayed response, This has been an unusually busy week

Thank you for the this bug report and for including all the necessary information required to reproduce the bug.  In this case Beagle's error message is intentional.  Beagle uses inter-marker correlation to phase and impute missing genotypes.  Consequently, Beagle's cannot phase or impute genotype if a marker window has only one marker.  

If your chromosome has more than one marker, you can eliminate the error message by increasing the window parameter to a value larger than the cM length of the longest chromosome (e.g. window=300 for human data), but this approach is recommended only if your data has clusters of tightly-linked markers separated by gaps without markers.  If your chromosome has only a single marker, the simplest solution is to exclude the marker from the analysis by including the marker's chromosome and position (in the format "CHROM:POS") in an excludemarkers file and using Beagle's excludemarkers= option.

Best regards,

Brian




On Mon, May 30, 2022 at 6:18 PM Seung-been Steven Lee <[sbstevenlee@gmail.com](mailto:sbstevenlee@gmail.com)> wrote:
Dear Professor Browning,

Thanks for developing/maintaining an awesome program. My name is Seung-been Lee and I go by Steven.

Today I'm writing to you to report a potential bug. More specifically, when the input VCF file contains only one variant in a given window, Beagle throws an error depending on the program version (v5.1 vs. v5.2 vs v5.4):

Exception in thread "main" java.lang.IllegalArgumentException: Window has only one position: CHROM=10 POS=94833639
at vcf.MarkerMap.meanSingleBaseGenDist(MarkerMap.java:98)
at phase.FixedPhaseData.markerMap(FixedPhaseData.java:170)
at phase.FixedPhaseData.<init>(FixedPhaseData.java:117)
at main.Main.phaseAndImpute(Main.java:140)
at main.Main.main(Main.java:110)

I'm attaching test files so you can reproduce the issue yourself (https://drive.google.com/drive/folders/1KqHhk02k_joVC2OY7zPOFWP00JxDoCRf?usp=sharing). Below are the commands used to produce result files.

v5.1 = beagle.18May20.d20.jar
v5.2 = beagle.28Jun21.220.jar
v5.4 = beagle.05May22.33a.jar

java -Xmx2g -jar beagle.18May20.d20.jar gt=data.vcf ref=CYP26A1.vcf.gz out=v5.1 impute=false > v5.1.out 2> v5.1.error
java -Xmx2g -jar beagle.28Jun21.220.jar gt=data.vcf ref=CYP26A1.vcf.gz out=v5.2 impute=false > v5.2.out 2> v5.2.error
java -Xmx2g -jar beagle.05May22.33a.jar gt=data.vcf ref=CYP26A1.vcf.gz out=v5.4 impute=false > v5.4.out 2> v5.4.error
 
Interestingly, it seems the problem was supposedly fixed in a previous version of Beagle (v.5.0). From the beagle5_release_notes file (https://faculty.washington.edu/browning/beagle/beagle5_release_notes):

```
Beagle 5.0 (15Sep18.8a0) release notes
=============================================
* Fixed a bug appearing when analysis window has only one target marker
```

Please let me know if I'm missing something here.

Steven

**Comment by sbslee** [comment]
As you can see above, I've contacted Professor Browning at the University of Washington, who is the author of Beagle. The tldr is, Beagle is supposed to raise an error when there is only one marker in a given window, including when there is only one marker in the input VCF. He also admitted that the current behavior of Beagle when there is only one marker in VCF is not ideal, but he told me there is no immediate plan to update Beagle for this.

Therefore, I decided to update PyPGx rather than waiting for Beagle to be updated. Basically, the `api.utils.estimate_phase_beagle` method will first detect whether or not the input VCF has only one marker. If so, the method will psuedophase the VCF -- i.e. haplotype phasing variants without any statistical analysis (`0/1` to `0|1`).

With this update, I was able to genotype CYP26A1 using your files (I attached the results to the reply):

```
$ pypgx run-ngs-pipeline \
CYP26A1 \
CYP26A1-pipeline \
--variants CYP26A1.ref.fixed.vcf.gz \
--do-not-plot-allele-fraction
Saved VcfFrame[Imported] to: CYP26A1-pipeline/imported-variants.zip
Saved VcfFrame[Phased] to: CYP26A1-pipeline/phased-variants.zip
Saved VcfFrame[Consolidated] to: CYP26A1-pipeline/consolidated-variants.zip
Saved SampleTable[Alleles] to: CYP26A1-pipeline/alleles.zip
Saved SampleTable[Genotypes] to: CYP26A1-pipeline/genotypes.zip
Saved SampleTable[Phenotypes] to: CYP26A1-pipeline/phenotypes.zip
Saved SampleTable[Results] to: CYP26A1-pipeline/results.zip
```

Note that this update is made in the `0.17.0-dev` branch. The official release of `0.17.0` will be made some time next month. If you want to try this update, you can install the dev version locally:

```
$ git clone https://github.com/sbslee/pypgx
$ cd pypgx
$ git checkout 0.17.0-dev
$ pip install .
```

Also note that while I was fixing the issue, I found another bug in one of the dependencies -- the `api.pyvcf.VcfFrame.pseudophase` method in the `fuc` pacakge. The bug was also fixed in the `0.35.0-dev` branch of that package. Therefore, if you're interested in running pypgx `0.17.0-dev` then you need to install fuc `0.35.0-dev` locally as well.

Please let me know if you have any questions.

**Comment by sbslee** [comment]
Closing this. Please feel free to re-open it if necessary.

**Comment by Wilsonjsjunior** [comment]
Hi, @sbslee!!!

I had install pypgx `0.17.0-dev` but now I got this:

```
Saved VcfFrame[Imported] to: CYP2D6-pipeline-testing/imported-variants.zip
Saved VcfFrame[Phased] to: CYP2D6-pipeline-testing/phased-variants.zip
/Users/wilsonjunior/miniconda3/envs/pypgx/lib/python3.9/site-packages/fuc/api/pyvcf.py:5286: FutureWarning: The default dtype for empty Series will be 'object' instead of 'float64' in a future version. Specify a dtype explicitly to silence this warning.
  df = self.df.sort_values(by=['CHROM', 'POS'],
Saved VcfFrame[Consolidated] to: CYP2D6-pipeline-testing/consolidated-variants.zip
Saved SampleTable[Alleles] to: CYP2D6-pipeline-testing/alleles.zip
Traceback (most recent call last):
  File "/Users/wilsonjunior/miniconda3/envs/pypgx/bin/pypgx", line 8, in <module>
    sys.exit(main())
  File "/Users/wilsonjunior/miniconda3/envs/pypgx/lib/python3.9/site-packages/pypgx/__main__.py", line 33, in main
    commands[args.command].main(args)
  File "/Users/wilsonjunior/miniconda3/envs/pypgx/lib/python3.9/site-packages/pypgx/cli/run_ngs_pipeline.py", line 159, in main
    pipeline.run_ngs_pipeline(
  File "/Users/wilsonjunior/miniconda3/envs/pypgx/lib/python3.9/site-packages/pypgx/api/pipeline.py", line 252, in run_ngs_pipeline
    plot.plot_vcf_allele_fraction(
  File "/Users/wilsonjunior/miniconda3/envs/pypgx/lib/python3.9/site-packages/pypgx/api/plot.py", line 396, in plot_vcf_allele_fraction
    ax1, ax2 = _plot_vcf_allele_fraction_one(
  File "/Users/wilsonjunior/miniconda3/envs/pypgx/lib/python3.9/site-packages/pypgx/api/plot.py", line 77, in _plot_vcf_allele_fraction_one
    imported_variants.data.plot_region(sample, ax=ax2, k='#AD_FRAC_REF', label='REF')
  File "/Users/wilsonjunior/miniconda3/envs/pypgx/lib/python3.9/site-packages/fuc/api/pyvcf.py", line 3143, in plot_region
    raise ValueError('VcfFrame is empty')
ValueError: VcfFrame is empty
````

Command:
```
pypgx run-ngs-pipeline CYP2D6 CYP2D6-pipeline-testing --variants grch38-variants.vcf.gz --depth-of-coverage grch38-depth-of-coverage.zip --control-statistics control-statistics-VDR.zip --force
```

Thanks!

**Comment by sbslee** [comment]
@Wilsonjsjunior,

Did you mean to post your question in #61?

In any case, it looks like your input files are GRCh38 but you are running the pipeline in GRCh37. Please try again using with '--assembly GRCh38'.

**Comment by hongsheng-lai** [comment]
Hi @sbslee, 
I just encountered this issue, and I'm using `0.20.0` now. I'm wondering if the psuedophase function you mentioned in the previous comment in `0.17.0-dev` is updated to versions after it. Or it might be other possible issues?
```
Exception in thread "main" java.lang.IllegalArgumentException: Window has only one position: CHROM=chr19 POS=39240564
        at vcf.MarkerMap.meanSingleBaseGenDist(MarkerMap.java:98)
        at phase.FixedPhaseData.markerMap(FixedPhaseData.java:170)
        at phase.FixedPhaseData.<init>(FixedPhaseData.java:117)
        at main.Main.phaseAndImpute(Main.java:140)
        at main.Main.main(Main.java:110)
Traceback (most recent call last):
  File "/usr/local/bin/pypgx", line 8, in <module>
    sys.exit(main())
  File "/usr/local/lib/python3.8/dist-packages/pypgx/__main__.py", line 33, in main
    commands[args.command].main(args)
  File "/usr/local/lib/python3.8/dist-packages/pypgx/cli/run_ngs_pipeline.py", line 159, in main
    pipeline.run_ngs_pipeline(
  File "/usr/local/lib/python3.8/dist-packages/pypgx/api/pipeline.py", line 239, in run_ngs_pipeline
    phased_variants = utils.estimate_phase_beagle(
  File "/usr/local/lib/python3.8/dist-packages/pypgx/api/utils.py", line 876, in estimate_phase_beagle
    subprocess.run(command, check=True, stdout=subprocess.DEVNULL)
  File "/usr/lib/python3.8/subprocess.py", line 516, in run
    raise CalledProcessError(retcode, process.args,
subprocess.CalledProcessError: Command '['java', '-Xmx2g', '-jar', '/usr/local/lib/python3.8/dist-packages/pypgx/api/beagle.22Jul22.46e.jar', 'gt=/tmp/tmpqm3iukuz/input.vcf', 'chrom=chr19:39240552-39248006', 'ref=/root/pypgx-bundle/1kgp/GRCh38/IFNL3.vcf.gz', 'out=/tmp/tmpqm3iukuz/output', 'impute=false']' returned non-zero exit status 1.
```

Also, my `imported-variants.zip` has two variants, not one, so shouldn't it be okay for Beagle?
```
#CHROM	POS	ID	REF	ALT	QUAL	FILTER	INFO	FORMAT	HG002
19	39240564	.	A	G	.	.	.	GT:AD:DP:AF	1/1:0,36:.:0.000,1.000
19	39245644	.	CAT	C,CATAT	.	.	.	GT:AD:DP:AF	1/2:4,12,8:.:0.167,0.500,0.333
```

**Comment by sbslee** [comment]
@hongsheng-lai,

It appears that you are working with GIAB data. Could you send me your input VCF file at sbstevenlee@gmail.com? If it's too big, the `imported-variants.zip` file is also fine. Also the exact command line you used, please.

Without looking at the input file, I'm just speculating here, but my guess is that even though there are two variants in VCF, given how they are fairly far from each other (~5kb), it's possible that they were put into two separate haplotype windows by Beagle. But I will need to test this myself to be sure.

**Comment by sbslee** [comment]
@hongsheng-lai,

Thank you for sending the files. I fixed the problem in the `0.21.0-dev` branch by updating the `api.utils.estimate_phase_beagle()` method to handle the Beagle error "Window has only one position". Basically, PyPGx will resort to the [phase-by-extension algorithm](https://pypgx.readthedocs.io/en/latest/faq.html#phase-by-extension-algorithm) instead of statistical phasing. You can access the development version:

```
$ git clone https://github.com/sbslee/pypgx
$ cd pypgx
$ git checkout 0.21.0-dev
$ pip install .
```

Thank you again for reporting the issue! Please let me know if you have any questions.
